xbout/region.py

```diff
@@ -0,0 +1,589 @@
+                    ylower_ind += ybndry
+
+                if self.connection_upper is None:
+                    self.ny -= ybndry
+                    print('check ny 3', self.ny, self.connection_upper, connect_upper)
+
+                    # used to calculate y-coordinate of upper side (self.yupper)
+                    yupper_ind -= ybndry
+
+            # calculate start and end coordinates
+            #####################################
+            xcoord = ds.metadata['bout_xdim']
+            ycoord = ds.metadata['bout_ydim']
+
+            # Note the global coordinates used here are defined so that they are zero at
+            # the boundaries of the grid (where the grid includes all boundary cells),
+            # not necessarily the physical boundaries because constant offsets do not
+            # matter, as long as these bounds are consistent with the global coordinates
+            # defined in apply_geometry (we will only use these coordinates for
+            # interpolation) and it is simplest to calculate them with cumsum().
+
+            # dx is constant in any particular region in the y-direction, so convert to a
+            # 1d array
+            # Note that this is not the same coordinate as the 'x' coordinate that is
+            # created by default from the x-index, as these values are set only for
+            # particular regions, so do not need to be consistent between different
+            # regions (e.g. core and PFR), so we are not forced to use just the index
+            # value here.
+            dx = ds['dx'].isel(**{ycoord: self.ylower_ind})
+            dx_cumsum = dx.cumsum()
+            self.xinner = dx_cumsum[xinner_ind] - dx[xinner_ind]
+            self.xouter = dx_cumsum[xouter_ind - 1] + dx[xouter_ind - 1]
+
+            # dy is constant in the x-direction, so convert to a 1d array
+            dy = ds['dy'].isel(**{xcoord: self.xinner_ind})
+            dy_cumsum = dy.cumsum()
+            self.ylower = dy_cumsum[ylower_ind] - dy[ylower_ind]
+            self.yupper = dy_cumsum[yupper_ind - 1]
+
+    def __repr__(self):
+        result = "<xbout.region.Region>\n"
+        for attr, val in vars(self).items():
+            result += "\t" + attr + "\t" + str(val) + "\n"
+        return result
+
+    def getSlices(self, mxg=0, myg=0):
+        """
+        Return x- and y-dimension slices that select this region from the global
+        DataArray.
+
+        Returns
+        -------
+        xslice, yslice : slice, slice
+        """
+        xi = self.xinner_ind
+        if self.connection_inner is not None:
+            xi -= mxg
+
+        xo = self.xouter_ind
+        if self.connection_outer is not None:
+            xi += mxg
+
+        yl = self.ylower_ind
+        if self.connection_lower is not None:
+            yl -= myg
+
+        yu = self.yupper_ind
+        if self.connection_upper is not None:
+            yu += myg
+
+        return slice(xi, xo), slice(yl, yu)
```

Comment:
Why not return a dictionary of slices? If it's of the form `{'x': slice(xi, xo), 'y': slice(yl, yu)}` then it can be passed directly into `.isel()` elsewhere